src/pipe-checksum.1
.\" ========================================================================
.\"    _   _ _ ____            ____          _____
.\"   | | | (_)  _ \ ___ _ __ / ___|___  _ _|_   _| __ __ _  ___ ___ _ __
.\"   | |_| | | |_) / _ \ '__| |   / _ \| '_ \| || '__/ _` |/ __/ _ \ '__|
.\"   |  _  | |  __/  __/ |  | |__| (_) | | | | || | | (_| | (_|  __/ |
.\"   |_| |_|_|_|   \___|_|   \____\___/|_| |_|_||_|  \__,_|\___\___|_|
.\"
.\"      ---  High-Performance Connectivity Tracer (HiPerConTracer)  ---
.\"                https://www.nntb.no/~dreibh/hipercontracer/
.\" ========================================================================
.\"
.\" High-Performance Connectivity Tracer (HiPerConTracer)
.\" Copyright (C) 2015-2026 by Thomas Dreibholz
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Contact: dreibh@simula.no
.\"
.\" ###### Setup ############################################################
.Dd September 7, 2026
.Dt pipe\-checksum 1
.Os pipe\-checksum
.\" ###### Name #############################################################
.Sh NAME
.Nm pipe\-checksum
.Nd Write Stdin to File and Create a File with the Checksum of the Data
.\" ###### Synopsis #########################################################
.Sh SYNOPSIS
.Nm pipe\-checksum Ar file
.Op Fl D Ar MD5|\%SHA1|\%SHA256|\%SHA384|\%SHA512|\%... |\% Fl \-digest Ar MD5|\%SHA1|\%SHA256|\%SHA384|\%SHA512|\%...
.Nm pipe\-checksum
.Op Fl h | Fl \-help
.Nm pipe\-checksum
.Op Fl v | Fl \-version
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm pipe\-checksum
writes all data from standard input into a given file "<file>". Furthermore, it
computes the checksum of the data and writes it to a new file "<file>.checksum".
The used message digest algorithm can be configured, it defaults to SHA256.
The checksum format is BSD\-style checksum format, as written for example by
the sha256sum tool. Therefore, any tool supporting this format can be used
for comparison.
.Pp
.\" ###### Arguments ########################################################
.Sh ARGUMENTS
The following arguments may be provided:
.Bl -tag -width indent
.It Ar file
Specifies the output file name.
.It Fl D Ar MD5|SHA1|SHA256|SHA384|SHA512|... | Fl \-digest Ar MD5|SHA1|SHA256|SHA384|SHA512|...
Sets the message digest. Any digest supported by the underlying libcrypto
library is possible. Calling pipe\-checksum with no options, or with an invalid digest, provides the list of all supported digests. Default: SHA256.
.It Fl h | Fl \-help
Prints command help.
.It Fl v | Fl \-version
Prints program version.
.El
.\" ###### Exit status ######################################################
.Sh EXIT STATUS
The
.Nm
tool exits with 0 on success, and >0 in case of an error.
.\" ###### Examples #########################################################
.Sh EXAMPLES
.Bl -tag -width indent
.It date | pipe\-checksum date.txt
.It sha256sum \-\-check date.txt.checksum
.It date | pipe\-checksum date.txt \-\-digest SHA384
.It dd if=/dev/urandom bs=12345 count=67 | hexdump | pipe\-checksum random.txt
.It dd if=/dev/urandom bs=12345 count=67 | pipe\-checksum random.txt \-\-digest SHA512
.It sha512sum \-\-check random.txt.checksum
.It pipe\-checksum \-\-version
.It pipe\-checksum \-h
.El
.\" ###### See also #########################################################
.Sh SEE ALSO
.Xr md5sum 1
.Xr sha1sum 1
.Xr sha256sum 1
.Xr sha512sum 1
.\" ###### Notes ############################################################
.Sh NOTES
This program is part of HiPerConTracer. The latest version of HiPerConTracer can be found on the HiPerConTracer Homepage at
.Lk https://www.nntb.no/\(tidreibh/hipercontracer/ "HiPerConTracer Homepage" .
.\" ###### Authors ##########################################################
.Sh AUTHORS
Thomas Dreibholz,
.Lk https://www.nntb.no/\(tidreibh "Homepage"


ChangeLog
commit c2586671eedee0068fc623622fe59f058b1e40c2
Merge: b00a9968 1c3000d6
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:59:49 2026 +0200

    Merge branch 'dreibh/homebrew' of github.com:dreibh/hipercontracer into dreibh/homebrew

commit b00a9968e3a706ba6c4b79e7fd74d51753cc3ba1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:58:41 2026 +0200

    Removed exit() call from library. DatabaseConfiguration::readConfiguration() returns false in case of a path error.

commit 1c3000d647dae82789040f95ff5a1b8b3853a65c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:55:16 2026 +0200

    Updated RPM spec file.

commit a2188f7125d634ff6e0d0ede00c527730536205e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:47:00 2026 +0200

    Using CMAKE_INSTALL_SYSCONFDIR for configuration file installation instead of hard-coding /etc.

commit 52ac9bc09bb4165923d0e3b2781497cab9d9f3fb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:30:42 2026 +0200

    Improved Homebrow path handling in CMakeLists.txt.

commit 2be6cc58880f261804db093b06920e4f13522ecb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:10:01 2026 +0200

    Some progress with the Homebrew port.

commit 21a740847032ad7e2b86da2571211a37e641d2bc
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 22:57:47 2026 +0200

    Some improvements.

commit f8a3f894156a3938dc44c7e3015ee4b50ab2142d
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 22:43:15 2026 +0200

    Initial homebrew packaging. Not working, yet!

commit 112761ee05490e780af98b8b80b29d4667c3ac86
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 17:07:58 2026 +0200

    Improved RPM spec file.

commit 4ab4e15001f36985d99663396bf93af46bd22056
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:57:57 2026 +0200

    Improved documentation.

commit 2f71eef68655e24b53083325c2bf7b6c569f3cd0
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:38:42 2026 +0200

    Setting capabilities (setcap cap_net_raw=ep) for Debian/Ubuntu and Fedora package.

commit 669311cc09b610fab6a00f170830bd9535d05980
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:37:25 2026 +0200

    Linux: Allow HiPerConTracer to run with capabilities (setcap cap_net_raw=ep) instead of startup with root permissions.

commit ef1049418431815e5674099d26b8d55f69770031
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:00:14 2026 +0200

    Another set of manpage improvements.

commit 3763e8cea0b34be3a7d0b76bc70b7e320f27f2f9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 15:54:43 2026 +0200

    Further manpage improvements.

commit b0f30cc37b8ad936594ea882a6a80d07f05d9a9c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 15:37:42 2026 +0200

    Updated manpages.

commit f2ddd54b04e55f370bf96964937d7586b93abcad
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 15:37:23 2026 +0200

    Use positional parameters as destinations.

commit b701a523bb3e3177283a1254527dfb63f121b216
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:22:15 2026 +0200

    Updated APKBUILD file.

commit cc74111202cfbc62d05470f40a9f4fd29b1b0d1c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:07:12 2026 +0200

    New release hipercontracer-2.2.10.

commit e2318e7c008132ecf89cb3391c9906a92d3962a7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 15:10:01 2026 +0200

    Fix for Debian test 30-dbeaver-tools.

commit 813403b8077be2520460d15674cd634399ec11db
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 16:23:32 2026 +0200

    Updated APKBUILD file.

commit 5c3c467cccde4b0e0fb78868650085f2fd890c94
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 15:08:02 2026 +0200

    Improved example scripts.

commit bb7d3cbfe79dc06f3c4b2a5fc3687cbb2d5b1c80
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:57:44 2026 +0200

    Updated script.

commit 25c5684be31fec6f4c923e94461509b4c71552ae
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:54:06 2026 +0200

    Added test script.

commit bb15335db526f34f8322d27270d63064eed58e3f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:51:51 2026 +0200

    Added Ping example script as well.

commit f2cdf7327f20c5328b9b075839ce913bd83a56de
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:46:11 2026 +0200

    Further clean-ups.

commit b29a86abd30967b176eb5266098e2e16039d24e4
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:39:31 2026 +0200

    Further improvements of the examples, and some function name clean-ups.

commit 14334a09f1948fa48c0648274b11b13839ef9da9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 12:53:25 2026 +0200

    Added another Python example.

commit 3269229394f112f1196d532d39992d2e344201a3
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 12:53:17 2026 +0200

    Improved Python library.

commit 3444d978211dfbab34ca19ced5eec21c350dd822
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 16:33:02 2026 +0200

    New release hipercontracer-2.2.9.

commit b9f4f55dab00295fc3c299777a4718b630170651
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 16:29:16 2026 +0200

    Version bump.

commit 1847ee70e030bc1c5fff9c9a0256b83af48a0ad2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 15:28:13 2026 +0200

    Updated test scripts.

commit 269f102483fea833bfd2aae28e8b9036ab180e89
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 14:51:17 2026 +0200

    Manpage fix.

commit cebc281513d0e5ee95bae5eec470356485305c1a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 14:49:31 2026 +0200

    Improved manpages.

commit c3cc031b9c8a3363ab2ddbc2af3a028163a1ee72
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Aug 11 15:43:18 2026 +0200


src/pipe-checksum.cc
// ==========================================================================
//     _   _ _ ____            ____          _____
//    | | | (_)  _ \ ___ _ __ / ___|___  _ _|_   _| __ __ _  ___ ___ _ __
//    | |_| | | |_) / _ \ '__| |   / _ \| '_ \| || '__/ _` |/ __/ _ \ '__|
//    |  _  | |  __/  __/ |  | |__| (_) | | | | || | | (_| | (_|  __/ |
//    |_| |_|_|_|   \___|_|   \____\___/|_| |_|_||_|  \__,_|\___\___|_|
//
//       ---  High-Performance Connectivity Tracer (HiPerConTracer)  ---
//                 https://www.nntb.no/~dreibh/hipercontracer/
// ==========================================================================
//
// High-Performance Connectivity Tracer (HiPerConTracer)
// Copyright (C) 2015-2026 by Thomas Dreibholz
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// Contact: dreibh@simula.no

#include "package-version.h"

#include <cstdio>
#include <cstdlib>
#include <fcntl.h>
#include <openssl/evp.h>

#include <chrono>
#include <string>
#include <cstring>
#include <iostream>
#include <boost/program_options.hpp>


// ###### Print digest algorithm ############################################
static void printDigest(const OBJ_NAME *obj, void* arg)
{
  std::cerr << obj->name << " ";
}


// ###### List all supported digest algorithms ##############################
static void listDigests()
{
   std::cerr << "Supported digests: ";
   OpenSSL_add_all_digests();
   OBJ_NAME_do_all(OBJ_NAME_TYPE_MD_METH, printDigest, nullptr);
   std::cerr << "\n";
}



// ###### Main program ######################################################
int main(int argc, char** argv)
{
   // ====== Handle arguments ===============================================
   if(argc < 2) {
      std::cerr << "Usage: " << argv[0] << " file [-D|--digest SHA256|...] [-h|--help] [-v|--version]\n";
      listDigests();
      return 1;
   }


   // ====== Initialize =====================================================
   std::string digestName("SHA256");
   std::string outputFileName;

   boost::program_options::options_description visibleOptions;
   visibleOptions.add_options()
      ( "help,h",
           "Print help message" )
      ( "version,v",
           "Show program version" )
      ( "digest,D",
           boost::program_options::value<std::string>(&digestName),
           "Digest" )
      ;
   boost::program_options::options_description hiddenOptions;
   hiddenOptions.add_options()
      ( "output,o",
        boost::program_options::value<std::string>(&outputFileName),
        "Output file" )
      ;
   boost::program_options::options_description commandLineOptions;
   commandLineOptions.add(visibleOptions);
   commandLineOptions.add(hiddenOptions);
   boost::program_options::positional_options_description positionalOptions;
   positionalOptions.add("output", 1);

   // ====== Handle command-line arguments ==================================
   boost::program_options::variables_map vm;
   try {
      boost::program_options::store(boost::program_options::command_line_parser(argc, argv).
                                       style(
                                          boost::program_options::command_line_style::style_t::unix_style
                                       ).
                                       options(commandLineOptions).
                                       positional(positionalOptions).
                                       run(), vm);
      boost::program_options::notify(vm);
   }
   catch(std::exception& e) {
      std::cerr << "ERROR: Bad parameter: " << e.what() << "\n";
      return 1;
   }

   if(vm.count("help")) {
      std::cerr << "Usage: " << argv[0] << " OPTIONS" << "\n"
                << commandLineOptions;
      return 0;
   }
   if(vm.count("version")) {
      std::cerr << "Pipe Checksum" << " " << HPCT_VERSION << "\n";
      return 0;
   }
   if(digestName == "") {
      listDigests();
      return 0;
   }
   const EVP_MD* md = EVP_get_digestbyname(digestName.c_str());
   if(md == nullptr) {
      std::cerr << "ERROR: Unknown message digest " << digestName << "!\n";
      listDigests();
      return 1;
   }
   if(outputFileName.empty()) {
      std::cerr << "ERROR: No output file given!\n";
      return 1;
   }


   // ====== Create output files ============================================
   const std::string                           checksumFileName(outputFileName + ".checksum");
   const std::string                           tmpOutputFileName(outputFileName + ".tmp");
   const std::string                           tmpChecksumFileName(checksumFileName + ".tmp");
   unsigned long long                          totalBytesWritten = 0;
   bool                                        success           = false;
   const std::chrono::system_clock::time_point t1                = std::chrono::system_clock::now();

   remove(outputFileName.c_str());
   remove(checksumFileName.c_str());
   FILE* outputFile = fopen(tmpOutputFileName.c_str(), "w");
   if(outputFile != nullptr) {
#ifdef POSIX_FADV_SEQUENTIAL
      if(posix_fadvise(fileno(outputFile), 0, 0, POSIX_FADV_SEQUENTIAL|POSIX_FADV_NOREUSE) < 0) {
         std::cerr << "WARNING: posix_fadvise() failed: " << strerror(errno);
      }
#endif

      FILE* checksumFile = fopen(tmpChecksumFileName.c_str(), "wt");
      if(checksumFile != nullptr) {
         EVP_MD_CTX* digestNameContext;
         digestNameContext = EVP_MD_CTX_create();
         if(digestNameContext != nullptr) {
            EVP_DigestInit_ex(digestNameContext, md, NULL);

            // ====== I/O loop ==============================================
            char              buffer[16384];
            ssize_t           length;
            while( (length = fread(&buffer, 1, sizeof(buffer), stdin)) > 0 ) {
               EVP_DigestUpdate(digestNameContext, buffer, length);
               if(fwrite(&buffer, length, 1, outputFile) != 1) {
                  std::cerr << "ERROR: Writing to " << tmpOutputFileName << " failed: " << strerror(errno) << "!\n";
                  goto finish;
               }
               totalBytesWritten += (unsigned long long)length;
            }

            // ====== Write checksum ========================================
            unsigned char mdValue[EVP_MAX_MD_SIZE];
            unsigned int  mdLength;
            EVP_DigestFinal_ex(digestNameContext, mdValue, &mdLength);
            fprintf(checksumFile, "%s (%s) = ", digestName.c_str(), outputFileName.c_str());
            for(unsigned int i = 0; i < mdLength; i++) {
               fprintf(checksumFile, "%02x", mdValue[i]);
            }
            fputs("\n", checksumFile);

            success = true;

finish:
            EVP_MD_CTX_destroy(digestNameContext);
         }
         else {
            std::cerr << "ERROR: EVP_MD_CTX_create() failed!\n";
         }
         EVP_cleanup();

         // ====== Close files ==============================================
         if(fclose(checksumFile)) {
            std::cerr << "ERROR: Unable to close checksum file " << tmpChecksumFileName << "!\n";
            success = false;
         }
      }
      if(fclose(outputFile)) {
         std::cerr << "ERROR: Unable to close output file " << tmpOutputFileName << "!\n";
         success = false;
      }
   }
   else {
      std::cerr << "ERROR: Unable to write output file " << tmpOutputFileName << "!\n";
   }


   // ====== Rename files ===================================================
   if(success) {
      if(rename(tmpOutputFileName.c_str(), outputFileName.c_str())) {
         std::cerr << "ERROR: Unable to rename " << tmpOutputFileName << " to " << outputFileName << "!\n";
         success = false;
      }
      if(rename(tmpChecksumFileName.c_str(), checksumFileName.c_str())) {
         std::cerr << "ERROR: Unable to rename " << tmpChecksumFileName << " to " << checksumFileName << "!\n";
         success = false;
      }
   }
   if(success) {
      const std::chrono::system_clock::time_point t2 = std::chrono::system_clock::now();

      // ====== Write statistics ============================================
      std::cerr << "Wrote " << totalBytesWritten << " B in "
                << std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1).count() << " ms => "
                << (totalBytesWritten / 1048576.0) / (std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count() / 1000000.0) << " MiB/s\n";
   }
   else {
      remove(tmpOutputFileName.c_str());
      remove(tmpChecksumFileName.c_str());
   }

   return (success == false);
}
